### PR TITLE
17 Write t-route simulated flow values to CHRTOUT files

### PR DESCRIPTION
**Pull request by awlostowski-noaa**
A helper function to write t-route simulated flows to WRF-Hydro CHRTOUT files. Results are passed from `compute_nhd_routing_SingleSeg.py` to a newly created function in `nhd_io.py` called  `write_q_to_wrf_hydro`. The following workflow steps happen in that function

1.  The stack of CHRTOUT files from which q-lateral data were extracted are opened using `xarray.open_mfdataset`. 
2. The `flowveldepth` data frame is reindexed to match the segment ordering in CHRTOUT files
3.  flow data are extracted from the `flowveldepth` data frame into a numpy array with the same dimensions as those within the CHRTOUT files. 
4. A new xarray data array object is created that contains t-route flow data with coordinates, dimensions, and attributes consistent with the CHTROUT data.
5. t-route flow data are added to the CHTROUT data object imported by `xarray.open_mfdataset`
6.  Break apart the CHTROUT stack data object into individual datasets, one for each export timestep
7. Save edited CHRTOUT files (containing t-route data) to disk.

### CODE REVIEW

**Comment by awlostowski-noaa** [comment]
I tried to numpy-ify the process handing the `results` array list to the function that writes flow data to CHRTOUT and unpacking/repacking data via numpy workflows. I ran into a speedbump whereby the number of segments simulated by t-route is fewer than the number of segments in listed in the CHRTOUT file.

- [ ] Why do the number of segments differ between t-route and CHRTOUT? Note - even with no masking, the number still differ. 
- [ ] Do we expect these numbers to differ in operational contexts? 
- [ ] If they do differ, would we write NaN values to segments that have no t-route simulation results?

File: src/python_framework_v02/troute/nhd_io.py [hunk 1]
```diff
@@ -218,6 +218,62 @@ def drop_all_coords(ds):
     return ds.reset_coords(drop=True)
 
 
+def write_q_to_wrf_hydro(results, flowveldepth, chrtout_files, qts_subdivisions, index_col="feature_id"):
+    
+    # open chrtout files as a single xarray dataset
+    chrtout = xr.open_mfdataset(
+        chrtout_files,
+        combine="by_coords"
+    )
```
Comment (jameshalgren) [inline src/python_framework_v02/troute/nhd_io.py:227]: invert this into a "with" block, i.e., 
```
with xr.open_mfdataset(
        chrtout_files,
        combine="by_coords"
    ) as chrtout
```

**Comment by jameshalgren** [comment]
For discussion, reservoir outflow is coded to the replaced segment: 
https://github.com/NOAA-OWP/t-route/blob/master/src/python_routing_v02/fast_reach/mc_reach.pyx#L842-L845

How do we handle that? (issues, need to address missing reach indexes; need to make output happen for reservoirs, ... )

**Comment by awlostowski-noaa** [comment]
*Loose End*: Writing individual netCDF files for each WRF Hydro timestep is pretty time intensive. Is there a more expedient way to do his?

**Comment by groutr** [comment]
@awlostowski-noaa is the use of mf_dataset your bottleneck?

File: src/python_framework_v02/troute/nhd_io.py [hunk 2]
```diff
@@ -218,6 +218,61 @@ def drop_all_coords(ds):
...
+                units = "m3 s-1",
+            ),
+        )
+
+        # add t-route flow variable to CHRTOUT stack
+        chrtout["streamflow_troute"] = qtrt_DataArray
+
+        # build a list of datasets, 1 for each timestep
+        dataset_list = []
+        grp_object = list(chrtout.groupby("time"))
```
Comment (groutr) [inline src/python_framework_v02/troute/nhd_io.py:265]: We don't need to realize this list here.

File: src/python_framework_v02/troute/nhd_io.py [hunk 3]
```diff
@@ -218,6 +218,61 @@ def drop_all_coords(ds):
...
+        
+    for i in range(len(grp_object)):
```
Comment (groutr) [inline src/python_framework_v02/troute/nhd_io.py:267]:
```python
for grp, vals in iter(grp_object):
    vals = vals.expand_dims({'time': grp})
    dataset_list.append(vals)
```

**Comment by awlostowski-noaa** [comment]
> @awlostowski-noaa is the use of mf_dataset your bottleneck? Have you profiled the code to see where the time is being spent? My guess would be that mf_dataset is a suboptimal approach to this particular problem.

@groutr The most time is spent reading and writing netCDF files with `xr.open_mfdataset` and `xr.save_mfdataset`, respectively. Indeed, `xr.save_mfdataset` takes 7.5x longer than `xr.open_mfdataset`. My hope was that mfdataset would allow me work much faster by using the Dask library. However, I don't think I am actually leveraging Dask to make mfdataset commands as fast as they could be. **Do you have experience using Dask?**

**Comment by awlostowski-noaa** [comment]
https://ncar.github.io/xdev/posts/writing-multiple-netcdf-files-in-parallel-with-xarray-and-dask/

This article describes basically the same process as we are trying to do here. However, they use the `dask.distributed` library. We can do the same, just need to understand what is happening.

**Comment by groutr** [comment]
@awlostowski-noaa Try passing in `coords='minimal'` and `data_vars='minimal'` to `open_mfdataset`.

**Comment by awlostowski-noaa** [comment]
```python
xr.save_mfdataset(datasets, paths = chrtout_files_new)
```
vs. 
```python
for i, dat in enumerate(datasets):
    dat.to_netcdf(chrtout_files_new[i])
```
These two methods are equivalent, if `dask.distributed` is not used. My intitial attempts at using Dask to expedite `save_mfdataset` (`clinet = Clinet()`) led to the following warnings and a much longer compute time than without. 
`distributed.nanny - WARNING - Restarting worker` ???

**Comment by groutr** [comment]
@awlostowski-noaa that is incorrect. You don't need dask-distributed here. Dask-distributed is for working on a cluster of machines.

What the documentation says for `save_mfdataset` is that if you don't use datasets with dask arrays, it is equivalent to calling `to_netcdf` on each dataset. If the datasets have dask arrays, it will use threading to attempt to parallelize the write.

**Comment by groutr** [comment]
@awlostowski-noaa Do the dask arrays in the datasets survive the groupby operation?

File: src/python_framework_v02/troute/nhd_io.py [hunk 4]
```diff
@@ -218,6 +218,70 @@ def drop_all_coords(ds):
...
+                description = "River Flow, t-route",
+                units = "m3 s-1",
+            ),
+        )
+
+        # add t-route DataArray to CHRTOUT dataset
+        chrtout["streamflow_troute"] = qtrt_DataArray
+
+        # group by time
+        grp_object = chrtout.groupby("time")
```
Comment (groutr) [inline src/python_framework_v02/troute/nhd_io.py:267]: If time is already one of the dimensions, do we really need a groupby here?

**Comment by awlostowski-noaa** [comment]
@groutr Here is one of the datasets after the groupby command. Dask arrays appear to  survive group by. Perhaps there is a scale issue? Writing 24 files with `save_mfdataset` may not show an advantage over sequential `save_dataset`, but maybe there is an advantage at 2400?
```
Dimensions:         (feature_id: 374, reference_time: 1)
Coordinates:
    latitude        (feature_id) float32 dask.array<chunksize=(374,), meta=np.ndarray>
    longitude       (feature_id) float32 dask.array<chunksize=(374,), meta=np.ndarray>
  * feature_id      (feature_id) int32 4186117 4186199 ... 4185689 4185589
  * reference_time  (reference_time) datetime64[ns] 2018-01-01T01:00:00
    time            datetime64[ns] 2018-01-01T02:00:00
Data variables:
    crs             |S1 b''
    order           (feature_id) int32 dask.array<chunksize=(374,), meta=np.ndarray>
    elevation       (feature_id) float32 dask.array<chunksize=(374,), meta=np.ndarray>
    streamflow      (feature_id) float32 dask.array<chunksize=(374,), meta=np.ndarray>
    q_lateral       (feature_id) float32 dask.array<chunksize=(374,), meta=np.ndarray>
    velocity        (feature_id) float32 dask.array<chunksize=(374,), meta=np.ndarray>
    qSfcLatRunoff   (feature_id) float32 dask.array<chunksize=(374,), meta=np.ndarray>
    qBucket         (feature_id) float32 dask.array<chunksize=(374,), meta=np.ndarray>
    qBtmVertRunoff  (feature_id) float32 dask.array<chunksize=(374,), meta=np.ndarray>
    qTroute         (feature_id) float32 dask.array<chunksize=(374,), meta=np.ndarray>
Attributes:
    featureType:                timeSeries
    proj4:                      +proj=lcc +units=m +a=6370000.0 +b=6370000.0 ...
    model_initialization_time:  2018-01-01_01:00:00
    station_dimension:          feature_id
    model_output_valid_time:    2018-01-01_02:00:00
    model_total_valid_times:    24
    stream_order_output:        1
    cdm_datatype:               Station
    Conventions:                CF-1.6
    code_version:               
    model_output_type:          channel_rt
    model_configuration:        retrospective
    dev_OVRTSWCRT:              1
    dev_NOAH_TIMESTEP:          3600
    dev_channel_only:           0
    dev_channelBucket_only:     0
    dev:                        dev_ prefix indicates development/
```

**Comment by groutr** [comment]
@awlostowski-noaa is the underlying disk a networked filesystem? I'm presuming you are writing to local storage.

File: src/python_framework_v02/troute/nhd_io.py (hunk 4 above)
Comment (jameshalgren) [reply src/python_framework_v02/troute/nhd_io.py:267]: Good question. I suspect that when we subject this write procedure to a full CONUS test dataset (only available on the Cheyenne HPC, at the moment) we will need to make a number of adjustments to address performance. I'm going to create an issue staging that HPC test (and associated refactor) and include the note that we should confirm the need (or not) of this `groupby`.

**Comment by jameshalgren** [comment]
Rebased to recent changes (including diffusive addition) and force pushed. Rebase was clean and did not change test results.

**Comment by jameshalgren** [comment]
Regarding performance of the mfdataset save, I did the following simple test which showed that both solutions we have tried are remarkably slow...
Here is the change:
``` diff
diff --git a/src/python_framework_v02/troute/nhd_io.py b/src/python_framework_v02/troute/nhd_io.py
index c12f114..3078e3e 100644
--- a/src/python_framework_v02/troute/nhd_io.py
+++ b/src/python_framework_v02/troute/nhd_io.py
@@ -265,13 +265,18 @@ def write_q_to_wrf_hydro(flowveldepth, chrtout_files, qts_subdivisions):
     chrtout_files_new = []
     chrtout_files_new[:] = [s + ".TRTE" for s in chrtout_files]

-    # mfdataset solution - can theoretically be parallelised via dask.distributed
-    xr.save_mfdataset(dataset_list, paths=chrtout_files_new)
+    save_mfd(dataset_list, chrtout_files_new)
+
+    save_serial(dataset_list, chrtout_files_new)

+def save_mfd(dataset_list, paths):
+    # mfdataset solution - can theoretically be parallelised via dask.distributed
+    xr.save_mfdataset(dataset_list, paths)

-#     # pure serial solution - saving for timing tests against mfdataset
-#     for i, dat in enumerate(dataset_list):
-#         dat.to_netcdf(chrtout_files_new[i])
+def save_serial(dataset_list, paths):
+    # pure serial solution - saving for timing tests against mfdataset
+    for i, dat in enumerate(dataset_list):
+        dat.to_netcdf(paths[i])
```
 
Here is the output from cprofile sorted by cumtime (see lines 10 `save_mfd` and 11 `save_serial` -- save_serial is slightly faster, but this is only a handful of tiny files and a handful of timesteps...)
```
   1         2    0.000    0.000   13.703    6.852 compute_nhd_routing_SingleSeg_v02.py:15(<module>)
   2    1271/1    0.034    0.000   13.703   13.703 {built-in method builtins.exec}
   3         1    0.003    0.003   12.092   12.092 compute_nhd_routing_SingleSeg_v02.py:1012(main)
   4         1    0.001    0.001    8.462    8.462 nhd_io.py:222(write_q_to_wrf_hydro)
   5        48    0.007    0.000    5.129    0.107 common.py:190(sync)
   6        48    0.005    0.000    5.122    0.107 core.py:807(store)
   7        48    0.003    0.000    4.378    0.091 api.py:991(to_netcdf)
   8         1    0.000    0.000    3.741    3.741 nhd_io.py:272(save_mfd)
   9         1    0.000    0.000    3.740    3.740 api.py:1138(save_mfdataset)
  10        24    0.001    0.000    3.530    0.147 dataset.py:1452(to_netcdf)
  11         1    0.000    0.000    3.530    3.530 nhd_io.py:276(save_serial)
  12       100    0.001    0.000    3.181    0.032 base.py:142(compute)
  13       100    0.002    0.000    3.180    0.032 base.py:382(compute)
  14       100    0.004    0.000    3.101    0.031 threaded.py:33(get)
  15       100    0.027    0.000    3.092    0.031 local.py:347(get_async)
  16         3    0.001    0.000    2.693    0.898 api.py:710(open_mfdataset)
  17         1    0.000    0.000    2.681    2.681 api.py:1235(<listcomp>)
  18      4416    0.004    0.000    2.601    0.001 local.py:132(queue_get)
  19      4425    0.023    0.000    2.598    0.001 queue.py:147(get)
  20      3195    0.015    0.000    2.549    0.001 threading.py:263(wait)
  21     21646    2.530    0.000    2.530    0.000 {method 'acquire' of '_thread.lock' objects}
  22        79    0.002    0.000    2.065    0.026 api.py:290(open_dataset)
  23       100    0.007    0.000    2.006    0.020 optimization.py:23(optimize)
  24         3    0.000    0.000    1.937    0.646 api.py:908(<listcomp>)
  25        57    0.001    0.000    1.874    0.033 __init__.py:1(<module>)
  26         2    0.002    0.001    1.838    0.919 nhd_io.py:148(get_ql_from_wrf_hydro_mf)
  27        79    0.002    0.000    1.720    0.022 api.py:445(maybe_decode_store)
  28  1265/112    0.008    0.000    1.707    0.015 <frozen importlib._bootstrap>:966(_find_and_load)
  29  1239/112    0.006    0.000    1.703    0.015 <frozen importlib._bootstrap>:936(_find_and_load_unlocked)
  30  1017/111    0.006    0.000    1.686    0.015 <frozen importlib._bootstrap>:651(_load_unlocked)
  31   877/114    0.003    0.000    1.669    0.015 <frozen importlib._bootstrap_external>:672(exec_module)
  32  1331/113    0.001    0.000    1.663    0.015 <frozen importlib._bootstrap>:211(_call_with_frames_removed)
  33       100    0.011    0.000    1.558    0.016 blockwise.py:506(optimize_blockwise)
  34       196    0.216    0.001    1.544    0.008 blockwise.py:554(_optimize_blockwise)
  35        48    0.001    0.000    1.520    0.032 api.py:1113(dump_to_store)
  36        48    0.000    0.000    1.475    0.031 common.py:262(store)
  37        79    0.003    0.000    1.321    0.017 conventions.py:513(decode_cf)
  38    509/68    0.001    0.000    1.235    0.018 {built-in method builtins.__import__}
  39 19501/17177    0.015    0.000    1.149    0.000 <frozen importlib._bootstrap>:997(_handle_fromlist)
  40         1    0.001    0.001    1.010    1.010 build_tests.py:161(parity_check)
```


... and percall
```
   1    1271/1    0.034    0.000   13.703   13.703 {built-in method builtins.exec}
   2         1    0.003    0.003   12.092   12.092 compute_nhd_routing_SingleSeg_v02.py:1012(main)
   3         1    0.001    0.001    8.462    8.462 nhd_io.py:222(write_q_to_wrf_hydro)
   4         2    0.000    0.000   13.703    6.852 compute_nhd_routing_SingleSeg_v02.py:15(<module>)
   5         1    0.000    0.000    3.741    3.741 nhd_io.py:272(save_mfd)
   6         1    0.000    0.000    3.740    3.740 api.py:1138(save_mfdataset)
   7         1    0.000    0.000    3.530    3.530 nhd_io.py:276(save_serial)
   8         1    0.000    0.000    2.681    2.681 api.py:1235(<listcomp>)
   9         1    0.001    0.001    1.010    1.010 build_tests.py:161(parity_check)
  10         1    0.001    0.001    0.977    0.977 compute_nhd_routing_SingleSeg_v02.py:314(compute_nhd_routing_v02)
  11         1    0.000    0.000    0.944    0.944 parallel.py:946(__call__)
  12         1    0.001    0.001    0.927    0.927 nhd_network_utilities_v02.py:507(build_qlateral_array)
  13         2    0.002    0.001    1.838    0.919 nhd_io.py:148(get_ql_from_wrf_hydro_mf)
  14         3    0.001    0.000    2.693    0.898 api.py:710(open_mfdataset)
  15         1    0.000    0.000    0.850    0.850 api.py:1227(<listcomp>)
  16         1    0.000    0.000    0.838    0.838 nhd_network_utilities_v02.py:1(<module>)
  17         1    0.000    0.000    0.837    0.837 nhd_io.py:1(<module>)
  18         1    0.000    0.000    0.684    0.684 testing.py:1(<module>)
  19         3    0.000    0.000    1.937    0.646 api.py:908(<listcomp>)
  20         1    0.000    0.000    0.385    0.385 dataarray.py:1(<module>)
  21         1    0.000    0.000    0.349    0.349 dataset_plot.py:1(<module>)
  22         1    0.000    0.000    0.346    0.346 alignment.py:1(<module>)
  23         1    0.000    0.000    0.342    0.342 variable.py:1(<module>)
  24         1    0.000    0.000    0.332    0.332 formatting_html.py:1(<module>)
  25         1    0.000    0.000    0.330    0.330 __init__.py:16(<module>)
  26         1    0.000    0.000    0.295    0.295 duck_array_ops.py:5(<module>)
  27         1    0.000    0.000    0.286    0.286 dask_array_compat.py:1(<module>)
  28         1    0.000    0.000    0.250    0.250 __init__.py:106(<module>)
  29         3    0.000    0.000    0.727    0.242 combine.py:478(combine_by_coords)

```

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py [hunk 5]
```diff
@@ -1287,6 +1287,19 @@ def main():
                 copy=False,
             )
 
+        if output_parameters.get("write_chrtout", None):
+            if forcing_parameters.get("qlat_input_folder", None):
```
Comment (jameshalgren) [inline src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py:1291]: This inter-dependency between chrtout write and the qlat_input method is appropriate for the I/O prototype.

**Pull request merged**